### DISCUSSION TITLE
WorkMail - EWS - OauthAuthentication

### DISCUSSION THREAD

**Discussion by tamili2i** [ecederstrand/exchangelib#1177] [Q&A]
Hi,
I am using ExchangeLib to fetch Mail from different mailboxes from WorkMail. I have enabled Impersonation Role to fetch mails. How to implement OAuth Authentication for the same?

**Comment by ecederstrand** [comment]
I don't know how OAuth towards AWS WorkMail works. Generally speaking, the simplest OAuth solution in exchangelib is to fetch an access token via some means outside of exchangelib and then create a credentials instance like this: `OAuth2Credentials(client_id=None, client_secret=None, access_token='blablabla')`.

**Reply by tamili2i** [reply]
Is there any sample code available for it? I didn't understand how to implement the same.

**Reply by ecederstrand** [reply]
This project does not provide any sample code for getting access tokens from AWS WorkMail, if that's what you're asking for.

**Reply by tamili2i** [reply]
How to manage OAuth2Credentials in exchange-lib?

**Reply by ecederstrand** [reply]
What do you mean by "manage"? The code I posted above shows how to create an `OAuth2Credentials` that you can use as the credentials object for your `Account` instance. I've seen reports that exchangelib works with AWS WorkMail, but be aware that WorkBail is a reverse-engineered version of the EWS protocol from Microsoft, so there may be some differences.

**Reply by tamili2i** [reply]
I am not sure how to get client_id and tenant id to create OAuth.

**Reply by ecederstrand** [reply]
I don't think you need them. In theory at least, the access token would be enough.

**Reply by tamili2i** [reply]
Okay Got it. Thanks for your explanation.
